### DISCUSSION TITLE
How to listen on both local and public IP (with systemd-resolved at 127.0.0.53)?

### DISCUSSION THREAD

**Discussion by Klaaktu** [DNSCrypt/dnscrypt-proxy#2425] [Q&A] [answered]
How to listen on both local and public IP?
The example use `[::]` but it's not possible when `systemd-resolved` is on 127.0.0.53. Uninstalling `systemd-resolved` might not be appropriate as it provides API for some services.
Is there a way to specify the listen address as 127.0.0.1 and the host's public IP? (Ideally not hard-coding the public IP.)

**Comment by lifenjoiner** [comment, marked as answer]
If `[::]` is instantiated to IPv4 `127.0.0.53` on your OS, what about `listen_addresses = ['127.0.0.1:53', '[::]:53']`?

Amend:
If `[::]:53` leads to quit. You must specify the public IP then: `listen_addresses = ['127.0.0.1:53', 'the-public-IP:53']`.

**Comment by aer0nix** [comment]
`listen_addresses = ['[::]:53']`

**Comment by o101010** [comment]
Heuuuu... As I know, [::] correspond to all ip in IPV6 stack, such as 0.0.0.0 for IPV4. It's an abbreviation for 0000:0000:0000:0000:0000:0000:0000:0000.
 So, what's the problem with it ?

**Reply by Klaaktu** [reply]
I probably had "port in use" when I tried because of systemd-resolved at 127.0.0.53:53. (I should have pasted the log...)

**Reply by o101010** [reply]
It depense of what adresse your proxy need to autorise.
Suppose your server use interface1 with 192.168.1.10 + FE80::10, interface2 with 203.0.120.195 + 2001:620::211:24FF:FE80:C12C.
- With something like `listen_addresses = ['0.0.0.0:53', '[::]:53']`, the proxy reply to all destination IP, IPv4 and IPv6, on all interfaces.
- With `listen_addresses = ['127.0.0.1:53', '[::1]:53']`, the proxy reply only to himself. Those IPs corresponds to *localhost*. 
- With `listen_addresses = ['192.168.1.10:53', '[FE80::10]:53']`, it only answer to people asking to those IPs (interface1). Not to localhost or interface2
- With `listen_addresses = ['192.168.1.10:53', '127.0.0.1:53]`, it reply to himself and other asking to 192.168.1.10. It can't respond to IPv6 connections.
- With `listen_addresses = ['127.0.0.1:53', '192.168.1.10:53', '[::1]:53', '[FE80::10]:53']`, it answer to yourself and people asking in interface1. Not on interface2
- With `listen_addresses = ['127.0.0.1:53', '192.168.1.10:53', '[::]:53']`, it's for yourself with IPv4, interface1 with IPv4 and everyone with IPv6 (loopback, interface1 and 2)
- ...

You can set as much IP addresses as you need. If you don't use IPv6, you don't need to set them on DNSCrypt.

If the port 53 was already used by another application, DNSCrypt-proxy just crash on starts.
